runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/ProcessBundleDescriptors.java

```diff
@@ -146,13 +189,89 @@ private static TargetEncoding addStageOutput(
         wireCoder);
   }
 
+  private static Map<String, Map<String, MultimapSideInputSpec>> forMultimapSideInputs(
+      ExecutableStage stage,
+      Components components,
+      ProcessBundleDescriptor.Builder bundleDescriptorBuilder) throws IOException {
+    ImmutableTable.Builder<String, String, MultimapSideInputSpec> idsToSpec =
+        ImmutableTable.builder();
+    for (SideInputReference sideInputReference : stage.getSideInputs()) {
+      // Update the coder specification for side inputs to be length prefixed so that the
+      // SDK and Runner agree on how to encode/decode the key, window, and values for multimap
+      // side inputs.
+      String pCollectionId = sideInputReference.collection().getId();
+      RunnerApi.MessageWithComponents lengthPrefixedSideInputCoder =
+          LengthPrefixUnknownCoders.forCoder(
+              components.getPcollectionsOrThrow(pCollectionId).getCoderId(),
+              components,
+              false);
+      String wireCoderId =
+          addWireCoder(sideInputReference.collection(), components, bundleDescriptorBuilder);
+      String lengthPrefixedSideInputCoderId = SyntheticComponents.uniqueId(
+          String.format(
+              "fn/side_input/%s",
+              components.getPcollectionsOrThrow(pCollectionId).getCoderId()),
+          bundleDescriptorBuilder.getCodersMap().keySet()::contains);
+
+      bundleDescriptorBuilder.putCoders(
+          lengthPrefixedSideInputCoderId, lengthPrefixedSideInputCoder.getCoder());
+      bundleDescriptorBuilder.putAllCoders(
+          lengthPrefixedSideInputCoder.getComponents().getCodersMap());
+      bundleDescriptorBuilder.putPcollections(
+          pCollectionId,
```

Comment:
I'm not sure this is what we want (though it should should not change correctness). What if this PCollection is consumed as a normal input to some transform here? We're mutating the coder used by the internal SDK nodes during bundle processing. Prior to this change, PCollections internally used their original coders and our synthetic wire coders were only used for grpc reads and writes. Could you add a note (or todo or bug) about this?

Reply:
The side input pcollection can never be part of the same executable stage as the one that is materializing it.

As a side note, the way that the coder is being mutated preserves SDK type safety so it would only impact PCollection size estimation.

Reply:
Sorry, I should have made that clearer. The same PCollection may be _consumed_ multiple times in the same executable stage: both as a side input and as a PCollection. Again, we maintain types here, so it should be "correct". It's just different from what we did before (where PCollection nodes that were actually processed _within_ an executable stage, as opposed to on the read nodes, used their original coders).

Reply:
PCollection which is consumed as a side input within an executable stage can only be consumed as a side input. You will never have an executable stage where the same PCollection is consumed as a main input and a side input.

Side input materialization and access requires a fusion break.

Reply:
Ah, interesting. I didn't realize those required separate materializations. Thanks for the explanation.

Reply:
It would be helpful to just put that as a comment into the code. I actually had the same doubt.